lisa/datautils.py

```diff
@@ -416,6 +416,90 @@ def merge(df1, df2):
     return functools.reduce(merge, df_list)
 
 
+def df_delta(pre_df, post_df, group_on):
+    """
+    pre_df and post_df containing paired/consecutive events indexed by time,
+    df_delta() merges the two dataframes and adds a ``delta`` column
+    containing the time spent between the two events.
+    A typical usecase would be adding pre/post events at the entry/exit of a
+    function.
+
+    Rows from ``pre_df`` and ``post_df`` are grouped by the ``group_on``
+    columns.
+    E.g.: ``['pid', 'comm']`` to group by task.
+    Except columns listed in ``group_on``, ``pre_df`` and ``post_df`` must
+    have columns with different names.
+
+    Events that cannot be paired are ignored.
+
+    :param pre_df: Dataframe containing the events that start a record.
+    :type pre_df: pandas.DataFrame
+
+    :param post_df: Dataframe containing the events that end a record.
+    :type post_df: pandas.DataFrame
+
+    :param group_on: Columns used to group ``pre_df`` and ``post_df``.
+        E.g.: This would be ``['pid', 'comm']`` to group by task.
+    :type group_on: list(str)
+
+    :returns: a :class:`pandas.DataFrame` indexed by the ``pre_df`` dataframe
+        with:
+
+        * All the columns from the ``pre_df`` dataframe.
+        * All the columns from the ``post_df`` dataframe.
+        * A ``delta`` column (duration between the emission of a 'pre' event
+            and its consecutive 'post' event).
+    """
+    pre_df = pre_df.copy(deep=False)
+    post_df = post_df.copy(deep=False)
+
+    # Tag the rows to remember from which df they are coming from.
+    pre_df["is_pre"] = True
+    post_df["is_pre"] = False
+
+    # Merge on columns common to the two dfs to avoid overlapping of names.
+    on_col = list(sorted(pre_df.columns & post_df.columns))
+
+    # Merging on nullable types converts columns to object.
+    # Merging on non-nullable types converts integer/boolean to float.
+    # Thus, let the on_col non-nullable and converts the others to nullable.
+    pre_df_cols = list(set(pre_df) - set(on_col))
+    post_df_cols = list(set(post_df) - set(on_col))
+    pre_df[pre_df_cols] = df_convert_to_nullable(pre_df[pre_df_cols])
+    post_df[post_df_cols] = df_convert_to_nullable(post_df[post_df_cols])
+
+    # Merge. Don't allow column renaming.
+    df = pd.merge(pre_df, post_df, left_index=True, right_index=True, on=on_col,
+                  how='outer', suffixes=(False, False))
+    df.index.name = 'Time'
```

Comment:
nit: we could save the current index name and reapply it afterwards. May be a good idea to ensure there is no conflict with existing columns by using a UUID as temp name: 
```
import uuid
index_col = uuid.uuid4().hex
```